src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -0,0 +1,595 @@
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Checks that a local variable is declared and/or assigned, but not used.
+ * </p>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name=&quot;UnusedLocalVariable&quot;/&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * public void method(int b) {
+ *     int a = 10; // violation
+ * }
+ * </pre>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code unused.local.var}
+ * </li>
+ * </ul>
+ *
+ * @since 9.2
+ */
+@FileStatefulCheck
+public class UnusedLocalVariableCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_UNUSED_LOCAL_VARIABLE = "unused.local.var";
+
+    /**
+     * An array of increment and decrement tokens.
+     */
+    private static final int[] INCREMENT_AND_DECREMENT_TOKENS = {
+        TokenTypes.POST_INC,
+        TokenTypes.POST_DEC,
+        TokenTypes.INC,
+        TokenTypes.DEC,
+    };
+
+    /**
+     * An array of scope tokens.
+     */
+    private static final int[] SCOPES = {
+        TokenTypes.SLIST,
+        TokenTypes.LITERAL_IF,
+        TokenTypes.LITERAL_FOR,
+        TokenTypes.LITERAL_WHILE,
+    };
+
+    /**
+     * An array of scope tokens which are exclusive to pattern variables.
+     */
+    private static final int[] SCOPES_EXCLUSIVE_TO_PATTERN_VARIABLES = {
+        TokenTypes.EXPR,
+        TokenTypes.ELIST,
+    };
+
+    /**
+     * Keeps VariableDesc objects inside stack of declared variables.
+     */
+    private Deque<VariableDesc> variables;
+
+    @Override
+    public int[] getDefaultTokens() {
+        return new int[] {
+            TokenTypes.DOT,
+            TokenTypes.VARIABLE_DEF,
+            TokenTypes.IDENT,
+            TokenTypes.SLIST,
+            TokenTypes.LITERAL_INSTANCEOF,
+            TokenTypes.EXPR,
+            TokenTypes.ELIST,
+            TokenTypes.LITERAL_IF,
+            TokenTypes.LITERAL_FOR,
+            TokenTypes.LITERAL_WHILE,
+        };
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public void beginTree(DetailAST root) {
+        variables = new ArrayDeque<>();
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        switch (ast.getType()) {
+            case TokenTypes.DOT:
+                final DetailAST firstDotOrMethodCallAst = getFirstDotOrMethodCallToken(ast);
+                if (firstDotOrMethodCallAst == null
+                        && ast.findFirstToken(TokenTypes.LITERAL_NEW) == null
```

Comment:
This can be simplified to
```java
                        && !TokenUtil.findFirstTokenByPredicate(ast, child -> {
                            return TokenUtil.isOfType(child,
                                    TokenTypes.LITERAL_NEW,
                                    TokenTypes.LITERAL_SUPER,
                                    TokenTypes.LITERAL_CLASS,
                                    TokenTypes.LITERAL_THIS);
                            }).isPresent()
```

Reply:
done